File: plugins/plugin-zalo/typescript/src/actions/sendMessage.ts
import type {
  Action,
  ActionExample,
  ActionResult,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { ZALO_SERVICE_NAME } from "../constants";
import type { ZaloService } from "../service";

export const SEND_MESSAGE_ACTION = "SEND_ZALO_MESSAGE";

export const sendMessageAction: Action = {
  name: SEND_MESSAGE_ACTION,
  similes: [
    "ZALO_SEND_MESSAGE",
    "ZALO_REPLY",
    "ZALO_MESSAGE",
    "SEND_ZALO",
    "REPLY_ZALO",
  ],
  description: "Send a message to a Zalo user",

  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
  ): Promise<boolean> => {
    const source = message.content?.source;
    return source === "zalo";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult | undefined> => {
    const zaloService = (await runtime.getService(ZALO_SERVICE_NAME)) as
      | ZaloService
      | undefined;

    if (!zaloService) {
      if (callback) {
        await callback({
          text: "Zalo service not available",
        });
      }
      return { success: false, error: "Zalo service not initialized" };
    }

    const responseText = state?.values?.response?.toString() || "";
    const userId = message.content?.userId || message.content?.chatId;

    if (!userId) {
      if (callback) {
        await callback({
          text: "No user ID available",
        });
      }
      return { success: false, error: "Missing user ID" };
    }

    try {
      const messageId = await zaloService.sendTextMessage(
        userId as string,
        responseText,
      );

      if (callback) {
        await callback({
          text: responseText,
          action: SEND_MESSAGE_ACTION,
        });
      }

      return {
        success: true,
        data: {
          action: SEND_MESSAGE_ACTION,
          userId,
          text: responseText,
          messageId,
        },
      };
    } catch (error) {
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      if (callback) {
        await callback({
          text: `Failed to send message: ${errorMessage}`,
        });
      }
      return { success: false, error: errorMessage };
    }
  },

  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Send a message to this Zalo chat",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll send a message to this chat now.",
          actions: [SEND_MESSAGE_ACTION],
        },
      },
    ],
  ] as ActionExample[][],
};

File: plugins/plugin-zalo/rust/src/actions/send_message.rs
//! Send message action implementation.

use async_trait::async_trait;
use serde_json::Value;

use super::{ActionContext, ZaloAction};
use crate::error::Result;

/// Action that builds a JSON payload for sending a Zalo message.
pub struct SendMessageAction;

#[async_trait]
impl ZaloAction for SendMessageAction {
    fn name(&self) -> &'static str {
        "SEND_ZALO_MESSAGE"
    }

    fn description(&self) -> &'static str {
        "Send a message to a Zalo user"
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context.message.get("source").and_then(|v| v.as_str());
        Ok(source == Some("zalo"))
    }

    async fn execute(&self, context: &ActionContext) -> Result<Value> {
        let response_text = context
            .state
            .get("response")
            .and_then(|v| v.get("text"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        Ok(serde_json::json!({
            "action": self.name(),
            "user_id": context.user_id,
            "text": response_text,
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_zalo_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "zalo",
                "text": "Hello"
            }),
            user_id: "12345".to_string(),
            state: serde_json::json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_non_zalo_source() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "telegram",
                "text": "Hello"
            }),
            user_id: "12345".to_string(),
            state: serde_json::json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_execute() {
        let action = SendMessageAction;

        let context = ActionContext {
            message: serde_json::json!({
                "source": "zalo",
                "text": "Hello"
            }),
            user_id: "12345".to_string(),
            state: serde_json::json!({
                "response": {
                    "text": "Hello from bot!"
                }
            }),
        };

        let result = action.execute(&context).await.unwrap();
        assert_eq!(result["action"], "SEND_ZALO_MESSAGE");
        assert_eq!(result["user_id"], "12345");
        assert_eq!(result["text"], "Hello from bot!");
    }
}
